### ISSUE TITLE
JIT: heuristics in optInvertWhileLoop may be overly conservative

### ISSUE THREAD

**Issue by JosephTremoulet** [dotnet/runtime#6569]
For example, the loop in method `Fit` from benchmark [Puzzle](https://github.com/dotnet/coreclr/blob/74ac12bf9966c5f0f6c58b72bbb99b08727c3d4e/tests/src/JIT/Performance/CodeQuality/BenchI/Puzzle/Puzzle.cs#L45-L59) gets rejected as too costly, causing loop hoisting etc. not to run on it, but it's a good candidate for such optimizations and Jit64 performed them on it:

``` c#
    private bool Fit(int i, int j)
    {
        for (int k = 0; k <= _pieceMax[i]; k++)
        {
            if (_p[i][k])
            {
                if (_puzzle[j + k])
                {
                    return false;
                }
            }
        }

        return true;
    }
```

There's already [special consideration](https://github.com/dotnet/coreclr/blob/74ac12bf9966c5f0f6c58b72bbb99b08727c3d4e/src/jit/optimizer.cpp#L3592-L3596) for loop conditions that involve shared static helpers; maybe similar treatment should be applied to loop conditions involving `arrLen` operators.

In this particular case, getting past the heuristic would enable hoisting the load of `_p` and the sign extension of `i` from the loop, but hoisting more of `_p[i]` (like the length of `_p`, the bounds-check, and/or the whole load) is blocked by dotnet/runtime#6554.

category:cq
theme:loop-opt
skill-level:expert
cost:medium

**Comment by AndyAyersMS** [comment]
See also notes/prototype in #780. Tried extending the cloning to cover multiple statements, etc.

**Comment by BruceForstall** [comment]
This was fixed by #50982
